### PR TITLE
HPCC-25574 Roxie UDP layer resend logic

### PR DESCRIPTION
**Pull request by richardkchapman**
Added logic to resend packets that have not been received
Removed hard limit on received packet queue
Added ability to request from more than one sender
Added code to track resends, duplicates etc

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x]  Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
Lots

### CODE REVIEW

File: roxie/udplib/udptrs.cpp [hunk 1]
```diff
@@ -79,15 +80,127 @@ using roxiemem::DataBuffer;
...
+        sequence_t seq = header->sendSeq;
+        header->pktSeq |= UDP_PACKET_RESENT;
+        if (!count)
+        {
+            first = seq;
+        }
+        else if (seq - first >= TRACKER_BITS)
+        {
+            // This shouldn't happen if we have steps in place to block ending new until we are sure old have been delivered.
+            UNIMPLEMENTED;
```
Comment (ghalliday) [inline roxie/udplib/udptrs.cpp:119]: should possibly be throwUnexpected()

File: roxie/udplib/udptrs.cpp [hunk 2]
```diff
@@ -79,15 +80,127 @@ using roxiemem::DataBuffer;
...
+        return first;
+    }
+    unsigned numActive() const
+    {
+        return count;
+    }
+    bool canRecord(unsigned seq) const
+    {
+        return (count==0 || seq - first < TRACKER_BITS);
+
```
Comment (ghalliday) [inline roxie/udplib/udptrs.cpp:190]: trivial: extra newline

File: roxie/udplib/udptrs.cpp [hunk 3]
```diff
@@ -138,52 +252,150 @@ class UdpReceiverEntry : public IUdpReceiverEntry
...
+                {
+                    maxPackets = TRACKER_BITS-inflight;
+                    if (udpTraceLevel>2 || maxPackets == 0)
+                        DBGLOG("Can't send more than %d new packets or we will overwrite unreceived packets (%u in flight, %u active %u resending now)", maxPackets, inflight, resendList->numActive(), resending);
+                    // Note that this may mean we can't send any packets, despite having asked for permission to do so
+                    // We will keep on asking.
+                }
+            }
+        }
+        if (udpTraceLevel>2 || toSend.size() == 1)
```
Comment (ghalliday) [inline roxie/udplib/udptrs.cpp:388]: Do you still want the last condition?

File: roxie/udplib/udptrs.cpp [hunk 4]
```diff
@@ -138,52 +252,150 @@ class UdpReceiverEntry : public IUdpReceiverEntry
...
+            DBGLOG("Resending %u packets", (unsigned) toSend.size());
+        while (maxPackets && packetsQueued.load(std::memory_order_relaxed))
         {
             DataBuffer *buffer = popQueuedData();
             if (!buffer)
                 break;  // Suggests data was aborted before we got to pop it
             UdpPacketHeader *header = (UdpPacketHeader*) buffer->data;
+            header->sendSeq = nextSendSequence++;
```
Comment (ghalliday) [inline roxie/udplib/udptrs.cpp:396]: minor: could be non atomic inc.

File: roxie/udplib/udptrr.cpp [hunk 5]
```diff
@@ -800,6 +810,29 @@ class CReceiveManager : implements IReceiveManager, public CInterface
...
+            E->Release();
+        }
+        catch (...)
+        {
+            IException *E = MakeStringException(ROXIE_INTERNAL_ERROR, "Unexpected exception caught in CPacketCollator::run");
+            EXCLOG(E);
+            E->Release();
+        }
+        if (msgColl)
+            msgColl->noteDuplicate(pktHdr->pktSeq & UDP_PACKET_RESENT);
```
Comment (ghalliday) [inline roxie/udplib/udptrr.cpp:834]: trivial, slightly better with (pktHdr->pktSeq & UDP_PACKET_RESENT) != 0
(avoiding implicit casting warning on windows?)

**Review by ghalliday — APPROVED** [review]
@richardkchapman a few very minor comments.

**Review by ghalliday — APPROVED** [review]
I think now ready for merging

**Comment by richardkchapman** [comment]
@mckellyln Any comments before we merge this?

**Comment by mckellyln** [comment]
I think its good.  I was looking at it all yesterday, nothing I could find that needs changing.

**Review by mckellyln — APPROVED** [review]
Looks good.

**Pull request merged**